src/main/thrift/parquet.thrift

```diff
@@ -966,6 +985,23 @@ struct ColumnIndex {
 
   /** A list containing the number of null values for each page **/
   5: optional list<i64> null_counts
+
+  /**
+   * A list of Boolean values to determine pages that contain only NaNs. Only
+   * present for columns of type FLOAT and DOUBLE. If true, all non-null
+   * values in a page are NaN. Writers are suggested to set the corresponding
+   * entries in min_values and max_values to NaN, so that all lists have the same
+   * length and contain valid values. If false, then either all values in the
+   * page are null or there is at least one non-null non-NaN value in the page.
+   * As readers are supposed to ignore all NaN values in bounds, legacy readers
+   * who do not consider nan_pages yet are still able to use the column index
+   * but are not able to skip only-NaN pages.
+   */
+  6: optional list<bool> nan_pages
```

Comment:
Is this necessary? We already know:
* the NaN count for each page (in `nan_counts`)
* the null count for each page (in `null_counts`)
* the number of rows for each page (from the OffsetIndex)

It seems this might be enough to infer whether a page is all-NaN (except perhaps if there are repetition levels?).

Reply:
That said, if we do need an additional list (because of repeated columns?), it might be more worthwhile to add an `optional list<i64> value_counts` instead, as it would then benefit all column types.

Reply:
Personally I think `optional list<i64> value_counts` is more common, but I think null already has `null_counts`, and `value_counts` might consume more bytes for every leaf column.

Reply:
Yes, number of rows in the offset index isn't enough due to repeated values.

Apart from this, the suggestions seem to turn a bit in circles now. Note that all suggestions in this thread were already mentioned in [my earlier post where I depicted our possible options for the column index](https://github.com/apache/parquet-format/pull/196#issuecomment-1486416762).

@pitrou what you mentioned was my Option 2. I personally would prefer this as it feels like a useful thing to have anyway. Having said that, others pointed rightfully out that it would cost a few bytes even for non float columns. The value might be valuable for other tasks as well. For example, it could be used to quickly check how many nested values are in a page. By having these values one could sum up the nested values per column chunk by adding up all the value counts. This is currently a value that cannot be optained at all through statistics; instead one has to decode pages and count. For example, the SQL query `SELECT count(*) FROM some_nested_column;` could be fully answered with such a value_counts field.

@wgtmac your proposal was my Option 1 and actually my initial proposal (see previous commit). Note that you [earlier](https://github.com/apache/parquet-format/pull/196#pullrequestreview-1362171450) actually were against writing NaNs and rather preferred the nan_pages approach:

> Personally speaking, apart from adding a nan_count to the statistics, I would go with the option 3: adding a nan_pages bool list to the column index. I am not in favor of writing any NaN to min/max bounds.

Is your argument that if we now need to write the NaNs anyway, that we should in this case just use them instead of adding nan_pages? I do agree that this would save the extra field and I personally see nothing wrong in doing this. Readers need to be able to detect NaN values anyway (to ignore them), so readers should be able to use the same logic to determin min=max=NaN <=> all values are NaN.

As mentioned in my previous post where I compared the three approaches, I am happy to implement any of them and I think all of them will fulfill the requirements. In my personal opinion, I like the current approach with nan_pages actually the least, as it seems redundant if we have to write NaN values anyway and I see no problem in using NaN values for the "all values NaN check".

I also like the option of adding a value_counts field to the column index of all columns. It feels like a useful and missing field (that is not subsumed by offset index row counts for nested columns) and I would love to add it as well and I feel the few extra bytes will be so negligible in contrast to the actual data that no-one will ever care. Also it would enable us to do the check for all values NaN the same way in page statistics and in the column index.

So we're back at the three options I proposed:

1. Drop nan_pages and use my initial approach of "min=max=NaN <=> all values are NaN" in the column index
2. Drop nan_pages and instead add value_counts so we can use value_counts-null_counts==nan_counts to determine whether all values are null. (My personal favorite)
3. Retain the current state and use `nan_pages`

@wgtmac @mapleFU @gszadovszky could we arrive at a consensus here? I'm happy to adapt my PR to any of the solutions. @gszadovszky you also haven't mentioned your favorite, yet (you just pointed out that we have to write some valid value).

Reply:
Personally I like (3), because I think parquet-format changes so slowly, adding a `value_count` or others in it will not be used for a long time. But others seems ok to me, maybe I can write a benchmark that will these bytes make PageIndex larger and decoding it slower.

Reply:
I would vote on (1) because it would not store redundant data. I think `nan_pages` is not necessary. Meanwhile, we have to take care of the [boundary_order](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L976) as well. Since no total ordering is defined `boundary_order` shall not be either `ASCENDING` or `DESCENDING` if there is any NaN page.

Reply:
+1 for (1) as I have explained in this comment: https://github.com/apache/parquet-format/pull/196#discussion_r1231982021

Reply:
I'm okay with both (1) and (2), even though (2) sounds more generally useful.

Reply:
Assume that a legacy reader has page index, and `min == max == NaN`, do need make sure that it will not prune it now? If not, (1) is ok for me, because it doesn't introduce any redudent data.

Reply:
I'm ok with (1), and I guess Java and Rust implementors should check that if they've prune page index without checking nan. @gszadovszky @pitrou do we need to:

1. check the backward capability for nan and pruning?
2. or just first check the parquet version is ok?
3. or regard the reader doesn't handling min-max nan as a bug?

Reply:
@mapleFU From just reading the spec, I don't think we should have a backward compatibility problem, as legacy readers are already compelled to ignore NaNs if they find them anywhere. Thus, a legacy reader would ignore the NaN it finds in the column index and just not filter that page.

Also note that regardless of whether we do (1), (2), or (3) [we basically **have to** write NaN into min and max](https://github.com/apache/parquet-format/pull/196#issuecomment-1491890773). We have to write a valid value and every value except NaN would simply be wrong, if a page contains only NaNs. The approaches would just differ in what we write **in addition**, so to a legacy reader that wouldn't read anything new fields, the three approaches would be equal.

Reply:
If that I'm ok with (1), thanks!

Reply:
@mapleFU, it seems to me that NaN is only checked for column indexes at the write path in parquet-mr. (In this case the column index will be invalid and won't be written to the file.) For the read path, though, there is no such check. It means that legacy readers can come to incorrect results using FLOAT/DOUBLE column indexes after we start writing NaN values. (Sorry for the late conclusion, I've thought this check was implemented for both directions.)
The only way I can think of for backward compatible NaN handling is to define a [ColumnOrder](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L863) for FP values that includes NaNs as well. In case of we would also add support to row-group level statistics with NaNs. parquet-mr currently skip all kinds of min/max statistics for columns with not supported column orders.

Reply:
So, maybe a new `ColumnOrder` enum would be added here?

Reply:
No, `ColumnOrder` is specified per column in the footer and it is universally valid for any min/max statistics. See [here](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1057) for details.

Reply:
`1: TypeDefinedOrder TYPE_ORDER;`

Emmm maybe my word is confusing, it means that, a new order should defined here to hint that?

Reply:
It would also require to rewrite a couple of parts in the spec to support a new `ColumnOrder`. Because we did not introduce any `ColumnOrder` since the idea was introduced it might require a vote as well. And we need to investigate the other implementations whether they already reads this value and handles a potentially "unknown" value there. parquet-mr does.

Reply:
> Since no total ordering is defined `boundary_order` shall not be either `ASCENDING` or `DESCENDING` if there is any NaN page.

Hmm. I am not theoretically against this (as is: the underlying concern is reasonable), but I'm worried that some corners of the Parquet format are more and more becoming a smattering of special cases that implementations must be extra careful to implement correctly.

That said, it should also be easy for an implementation to entirely ignore `boundary_order`, and instead detect any existing ordering from the `min_values` and `max_values` (this should be fast given that there is one value per page). It might even be useful to deprecate `boundary_order` and encourage implementations to derive the information themselves?

Reply:
What is a good path forward then? I see the following options:

a) Ship this change but exclude the handling of only-Nan pages in the column index and only handle the other cases. Then we could still at least specify how to handle NaNs in the column index in cases where no "only NaN" page exists and these cases would then at least be well defined (only NaN pages are probably an edge case, so this would already allow us to filter in 99% of all cases and therefore get us almost to the goal).
b) Add ColumnOrder to this proposal. (again happy to do that)
It would be a good case to start using the ColumnOrder enum. This would also give us the opportunity to define `boundary_order` explicitly for this column order, so we could even assume an ordering.
c) Drop this altogether and live with the fact that float / double columns are basically unfilterable in many cases.

@gszadovszky 
Side note: I think that the current read behavior in parquet-mr as you state it is not adhering to the spec and is dangerous at best. I have seen Parquet files which have NaN in these bounds in the wild (I don't know who wrote them) and since the mandate to not write NaNs to these bounds is in the spec only for a while ([introduced here](https://github.com/apache/parquet-format/commit/92ae9a3187d7673c9a40f81f40886faa20807722)), older writers would have been perfectly spec-conforming when writing NaN into these bounds, so files having NaNs here are adhering to (an older version of the) spec and therefore the parquet-mr read code should be robust to handle these cases.

Reply:
Oh, actually there is yet another option

d) Stick with nan_pages (or value_counts) (i.e., alternatives (2) or (3)) and write min=-Infinity and max=+Infinity into the bounds in the column index for only-NaN pages. This way, new readers could use nan_pages (or value_counts) to detect an only-NaN pages. Legacy readers would simply never filter this page due to the maximally wide bounds. My heart is bleeding a bit while writing this, as this is obviously a patch solution that feels wrong (the bounds are just not correct) and is just to reverse-patch old implementations by bending the spec, but it would fulfill the requirements and allow backward compatibility while enabling support for filtering only-NaN pages.

Reply:
> d) Stick with nan_pages (or value_counts) (i.e., alternatives (2) or (3)) and write min=-Infinity and max=+Infinity into the bounds in the column index for only-NaN pages.

What if a page contains actual infinity values?

Reply:
@pitrou That shouldn't be a problem. That's why this approach would require alternative (2) or (3). In these alternatives nan_pages / value_counts would be used to find only-NaN pages. If these indicate that the page is only NaN, the min/max can be ignored and a reader can assume that the only values in the page are NaNs. Old readers who don't understand these new fields yet would treat the page simply as "maximum value range; cannot filter".

Reply:
@JFinis, your idea sounds good but it is not that easy, unfortunately. Since no total ordering is specified NaN values can get before negative infinity or after positive infinity. An implementation that currently writes NaN values to column indexes will break in this scenario.
@pitrou, I've brought up boundary order because that was our original answer to the problems of these ordering issues. NaN values are not the only potential issues around ordering. E.g. how should we order internationalized UTF-8 strings?
I agree that the current parquet-mr implementation of handling NaN values in column indexes is not correct. But it also means we cannot do this change without breaking older parquet-mr readers. Boundary order would solve this from parquet-mr point of view but if it is not used by other implementations it is not a good choice on its own either.

If there are parquet files with column indexes containing NaN values and we consider them valid then we need to fix this issue in parquet-mr and it is unrelated to this format change. However, it is not an easy question if they are really valid. Are both min and max are NaN? If not what is the total ordering in that system which writes these files? Can this format change be compatible with that system?

Reply:
@pitrou, sorry `BoundaryOrder` was a mistype. I was talking about [ColumnOrder](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L863). The only order we have currently is `TypeDefinedOrder` that is specified. We were thinking about adding a `ColumnOrder` for FLOAT/DOUBLE with the definition of a total ordering that includes NaN values, -0.0, 
and +0.0 values.
Maybe you're right that the in case of the default string ordering is not enough to a system it shall write its own indices. But there was an idea behind ColumnOrder to maybe implement collations to support those systems.

Reply:
I think `[-Inf, +Inf]` is ok.

> Add in comments that in the page index, all nan pages can be checked by having nan_count > 0 && min is NaN && max is NaN

Previous design uses `[Nan, Nan]`, I guess it's bad. But i guess `[-Inf,+inf]` should be well handled and not including any ambiguity. I'm +1 with this

Reply:
@JFinis, there was an argument that some writers already write NaN values into column indexes. Hence, they try to filter on NaN values. Now, we start writing `[-Inf,+Inf]` for NaN only pages. NaN is probably out of `[-Inf,+Inf]` interval so that reader would drop the only NaN page while searching for a NaN.

Reply:
> Now, we start writing `[-Inf,+Inf]` for NaN only pages.

Also, how does the reader distinguish with pages that contain actual infinity values?

Reply:
@gszadovszky Any writer reader/writer pair who writes NaNs into column indexes (and other places like page headers) and expects them to be there (and otherwise yields wrong results while reading) *is and never was* spec conforming. In older releases of the spec where NaN wasn't mentioned yet, such a writer was at least not violating the spec directly but even then NaN handling was basically "undefined behavior", as the spec never mentioned how to treat NaNs. Thus, relying on *one specific* behavior w.r.t. NaNs was already back then a non-portable assumption.

Even today, a reader relying on one specific NaN semantics would already yield erroneous results when reading spec conforming Parquet files. E.g., if they search for NaNs and expect them to be in min/max, then they might filter Pages containing NaNs that don't have NaNs in their min/max. Consequently, such a reader is already broken; yes, writing [-Inf,Inf] into the column index would break such a reader more, but all bets are off here anyway already. It currently is just not possible to handle NaNs correctly in a portable way (that's what this PR is all about in the first place).

So TBH backward compatibility to such a broken (or at least non-portable) reader/writer pair seems like an absolute non-goal to me.

@pitrou A legacy reader who doesn't handle the new NaN semantics doesn't need to distinguish here. All they need to know is whether they should skip the page or shouldn't. A page with [-Inf,+Inf] can never be skipped, so regardless of whether the bounds are there due to NaNs or real infinities, a legacy reader would yield correct results. A new reader that implements this PR can do the distinction via the nan_pages or value_counts computation.

Note that actually *any* bounds are, mathematically speaking, correct for a page containing only NaNs (and will yield correct results on spec-conforming readers!). Note that min/max values in the column index don't need to be tight, according to the spec. So the only condition that must hold is that there is no value outside of the bounds (NaNs excluded). As an only-NaN page has no values, any bounds satisfy the condition, as there are no values that need to lie inside them. So instead of [-Inf,+Inf] we could also choose [0,0] or [42,1337]. Both would yield correct results on spec conforming readers. Actually the tighter the bounds, the more queries can skip the page.

Reply:
> A new reader that implements this PR can do the distinction via the nan_pages or value_counts computation.

Wait... I thought the `[-Inf, +Inf]` convention was meant to avoid a new `nan_pages` or `value_counts` field? If not, then what's the point?

Reply:
@pitrou writing anything but NaN into min/max was my suggestion to circumvent the problem that [parquet-mr doesn't seem to check for NaN values in min/max while reading](https://github.com/apache/parquet-format/pull/196#discussion_r1243234931) and therefore would probably yield wrong results once we start writing NaNs into these values.

This would only work if we go back to maintaining either `nan_pages` or `value_counts` though, as otherwise, as you correctly pointed out, we wouldn't have a way to draw the distinction between only-NaN and real infinities.

Reply:
> there was an argument that some writers already write NaN values into column indexes. Hence, they try to filter on NaN values. Now, we start writing `[-Inf,+Inf]` for NaN only pages. NaN is probably out of `[-Inf,+Inf]` interval so that reader would drop the only NaN page while searching for a NaN.

Hi gabor, which implemention has do like that? I check C++ implemention but it doesn't do this. Maybe we can do a check here? Since I guess `[-inf, +inf]` sounds ok

Reply:
I think `[-inf, +inf]` it's ok. Now I guess only Rust impl and parquet-mr has the potential problem.

Reply:
Trying to catching up the discussion. I like the idea to write either [-inf, +inf] or [-0.0, +0.0] for NaN-only pages.

As NaN value does not have a well-defined order across systems, simply leveraging page min/max values to filter NaN does not make any sense. Therefore I think this design can break such misuses.